### ISSUE TITLE
tarfile not re-entrant for multi-threading

### ISSUE THREAD

**Issue by 006f2f76-b3a8-4e01-887f-8a098c3a1ee7** [python/cpython#67837]
BPO | [23649](https://bugs.python.org/issue23649)
--- | :---
Nosy | @gustaebel, @bitdancer, @xdegaye, @websurfer5
Files | <li>[mutithreading_tarfile.patch](https://bugs.python.org/file38462/mutithreading_tarfile.patch "Uploaded as text/plain at 2015-03-12.21:59:53 by sgnn7"): makdirs_patch</li><li>[extract_from_packages.py](https://bugs.python.org/file47523/extract_from_packages.py "Uploaded as text/plain at 2018-04-07.10:01:01 by @xdegaye")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2015-03-12.16:40:01.811>
labels = ['type-bug']
title = 'tarfile not re-entrant for multi-threading'
updated_at = <Date 2019-06-10.14:55:43.587>
user = 'https://bugs.python.org/sgnn7'
```

bugs.python.org fields:
```python
activity = <Date 2019-06-10.14:55:43.587>
actor = 'vstinner'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = []
creation = <Date 2015-03-12.16:40:01.811>
creator = 'sgnn7'
dependencies = []
files = ['38462', '47523']
hgrepos = []
issue_num = 23649
keywords = ['patch']
message_count = 11.0
messages = ['237960', '237961', '237964', '237968', '237970', '237986', '237989', '237990', '238181', '238183', '315067']
nosy_count = 5.0
nosy_names = ['lars.gustaebel', 'r.david.murray', 'xdegaye', 'sgnn7', 'Jeffrey.Kintscher']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue23649'
versions = ['Python 3.4']
```

</p></details>

### Linked PRs
* gh-103263
* gh-115082

**Comment by piskvorky** [comment]
I was also bitten by this. I'm processing large nested archives (TARs within TARs) in read-only mode in parallel, and noticed strange inconsistencies in the number of extracted files. No exceptions – just `for tarinfo in tar_file: …` iteration sometimes ended prematurely. I tracked this problem down to `tarfile` and multithreading.

I have to disagree with @gustaebel 's conclusion:

>  agree with David that there is no need for tarfile to be thread-safe. There is nothing to be gained from distributing one TarFile object among multiple threads because it operates on a single resource which has to be accessed sequentially anyway.

There is much to be gained by threading (at least in my application), and the resource does *not* have to be accessed sequentially. `tarfile` already seeks around, each object/method knows its offset.

Like @sgnn7, I fixed the issue by adding locks around TarFile's `open()` & `next()` & `extractfile()` calls, that did it. But not having to worry about this would be great. Actually noticing this issue in the first place was the biggest challenge – there were no exceptions! Just mysteriously incomplete data.

Unlike @sgnn7, I didn't get any exception – without the locks, `tarfile.next()` just silently swallows a `tarfile.InvalidHeaderError: invalid header` exception (caused by some internal seek/tell offset corruption due to multithreading), which means `next()` returned `tarinfo = None`, which means it ended the iteration early. No errors, just missing data.

As an aside, not having to call `tar_file.members = []` regularly during iteration (to avoid OOM due to TarFile's internal caching) would be great too. But that's not related to threading.
